api/v3/Contribution.php

```diff
@@ -616,84 +616,76 @@ function _civicrm_api3_contribution_completetransaction_spec(&$params) {
  * @throws API_Exception
  */
 function civicrm_api3_contribution_repeattransaction($params) {
+
   civicrm_api3_verify_one_mandatory($params, NULL, ['contribution_recur_id', 'original_contribution_id']);
+
+  // We need a contribution to copy.
   if (empty($params['original_contribution_id'])) {
+    // Find one from the given recur. A template contribution is preferred, otherwise use the latest one added.
     $templateContribution = CRM_Contribute_BAO_ContributionRecur::getTemplateContribution($params['contribution_recur_id']);
     if (empty($templateContribution)) {
       throw new CiviCRM_API3_Exception('Contribution.repeattransaction failed to get original_contribution_id for recur with ID: ' . $params['contribution_recur_id']);
     }
-    $params['original_contribution_id'] = $templateContribution['id'];
-  }
-  $contribution = new CRM_Contribute_BAO_Contribution();
-  $contribution->id = $params['original_contribution_id'];
-  if (!$contribution->find(TRUE)) {
-    throw new API_Exception(
-      'A valid original contribution ID is required', 'invalid_data');
   }
-  // We don't support repeattransaction without a related recurring contribution.
-  if (empty($contribution->contribution_recur_id)) {
-    throw new API_Exception(
-      'Repeattransaction API can only be used in the context of contributions that have a contribution_recur_id.',
-      'invalid_data'
-    );
+  else {
+    // A template/original contribution was specified by the params. Load it.
+    $templateContribution = Contribution::get(FALSE)
+      ->addWhere('id', '=', $params['original_contribution_id'])
+      ->addWhere('is_test', 'IN', [0, 1])
+      ->addWhere('contribution_recur_id', 'IS NOT EMPTY')
+      ->execute()->first();
+    if (empty($templateContribution)) {
+      throw new CiviCRM_API3_Exception("Contribution.repeattransaction failed to load the given original_contribution_id ($params[original_contribution_id]) because it does not exist, or because it does not belong to a recurring contribution");
+    }
   }
 
-  $params['payment_processor_id'] = civicrm_api3('contributionRecur', 'getvalue', [
-    'return' => 'payment_processor_id',
-    'id' => $contribution->contribution_recur_id,
-  ]);
-
-  $passThroughParams = [
+  // Collect inputs for CRM_Contribute_BAO_Contribution::completeOrder in $input.
+  $paramsToCopy = [
     'trxn_id',
-    'total_amount',
     'campaign_id',
     'fee_amount',
     'financial_type_id',
     'contribution_status_id',
     'payment_processor_id',
-  ];
-  $input = array_intersect_key($params, array_fill_keys($passThroughParams, NULL));
+    'is_email_receipt',
+    'trxn_date',
+    'receive_date',
+    'card_type_id',
+    'pan_truncation',
+    'payment_instrument_id',
+    // Q. not contribution_page_id ?
```

Comment:
I think we want to permit as few params as possible & add params (with tests) if there are use cases for them